### PR TITLE
feat(fac): Add custom TTL options for App Check

### PR DESCRIPTION
**Pull request by lahirumaramba**
- Add new `AppCheckTokenOptions` type
- Change the public API `createToken(...)` to accept optional `AppCheckTokenOptions`
- Add validations for custom `ttl`
- Move `transformMillisecondsToSecondsString()` to `utils`
- Add unit tests. ~(Integration tests will follow in a separate PR)~
- Add integration tests.

RELEASE NOTE: The `createToken()` API now supports configuring the `TTL` of the returned Firebase App Check Token.

### CODE REVIEW

**Review by weixifan — APPROVED** [review]

File: src/app-check/index.ts [hunk 1]
```diff
@@ -95,6 +96,18 @@ export namespace appCheck {
     ttlMillis: number;
   }
 
+  /**
+   * Interface representing an App Check token options.
```
Comment (hiranya911) [inline src/app-check/index.ts:100]: Remove "an"?

File: src/app-check/token-generator.ts [hunk 2]
```diff
@@ -98,6 +109,35 @@ export class AppCheckTokenGenerator {
...
+   * @returns A custom object with ttl converted to protobuf Duration string format.
+   */
+  private validateTokenOptions(options: AppCheckTokenOptions): {[key: string]: any} {
+    if (!validator.isNonNullObject(options)) {
+      throw new FirebaseAppCheckError(
+        'invalid-argument',
+        'AppCheckTokenOptions must be a non-null object.');
+    }
+    if (typeof options.ttlMillis !== 'undefined') {
+      if (!validator.isNumber(options.ttlMillis) || options.ttlMillis < 0) {
```
Comment (hiranya911) [inline src/app-check/token-generator.ts:127]: Negative check is redundant due to the following check.

File: test/unit/app-check/token-generator.spec.ts [hunk 3]
```diff
@@ -122,11 +122,49 @@ describe('AppCheckTokenGenerator', () => {
...
+
+    it('should be fulfilled with a Firebase Custom JWT with only an APP ID', () => {
       return tokenGenerator.createCustomToken(APP_ID)
         .should.eventually.be.a('string').and.not.be.empty;
     });
 
+    [THIRTY_MIN_IN_MS, THIRTY_MIN_IN_MS + 1, SEVEN_DAYS_IN_MS / 2, SEVEN_DAYS_IN_MS - 1, SEVEN_DAYS_IN_MS]
+      .forEach((ttlMillis) => {
+        it('should be fulfilled with a Firebase Custom JWT with a valid custom ttl' + JSON.stringify(ttlMillis), () => {
+          return tokenGenerator.createCustomToken(APP_ID, { ttlMillis })
```
Comment (hiranya911) [inline test/unit/app-check/token-generator.spec.ts:163]: We should probably decode the token and verify the expected TTL is set.

**Review by hiranya911 — APPROVED** [review]
LGTM with some suggestions.

File: src/app-check/token-generator.ts [hunk 4]
```diff
@@ -83,6 +93,7 @@ export class AppCheckTokenGenerator {
         aud: FIREBASE_APP_CHECK_AUDIENCE,
         exp: iat + ONE_HOUR_IN_SECONDS,
         iat,
+        ...customOptions,
```
Comment (hiranya911) [inline src/app-check/token-generator.ts:96]: Just curious. Why is `ttl` separate from `exp`?

Comment (lahirumaramba) [reply src/app-check/token-generator.ts:96]: According to go/fac-configurable-ttls we are not exposing the option to use `exp` at all to keep the interface simple. So in this case a custom `ttl` will override `exp`.

File: src/app-check/token-generator.ts (hunk 2 above)
Comment (lahirumaramba) [reply src/app-check/token-generator.ts:127]: I added the negative check here to specify that the ttl should be a non-negative duration in the error message. I agree though that it makes more sense to include/check that as part of the range validation below. Will update the code.

File: test/unit/app-check/token-generator.spec.ts (hunk 3 above)
Comment (lahirumaramba) [reply test/unit/app-check/token-generator.spec.ts:163]: Good call. Thanks!

**Comment by lahirumaramba** [comment]
Thanks! Updated the code with PR fixes.

Adding @kevinthecheung to review the reference docs. Thank you!

File: test/unit/app-check/token-generator.spec.ts [hunk 5]
```diff
@@ -157,11 +157,16 @@ describe('AppCheckTokenGenerator', () => {
         .should.eventually.be.a('string').and.not.be.empty;
     });
 
-    [THIRTY_MIN_IN_MS, THIRTY_MIN_IN_MS + 1, SEVEN_DAYS_IN_MS / 2, SEVEN_DAYS_IN_MS - 1, SEVEN_DAYS_IN_MS]
-      .forEach((ttlMillis) => {
-        it('should be fulfilled with a Firebase Custom JWT with a valid custom ttl' + JSON.stringify(ttlMillis), () => {
-          return tokenGenerator.createCustomToken(APP_ID, { ttlMillis })
-            .should.eventually.be.a('string').and.not.be.empty;
+    [[THIRTY_MIN_IN_MS, '1800s'], [THIRTY_MIN_IN_MS + 1, '1800.001000000s'],
```
Comment (hiranya911) [inline test/unit/app-check/token-generator.spec.ts:160]: One entry per line for clarity:

```
[
  [],
  [],
  ...
]
```

**Review by hiranya911 — APPROVED** [review]
Thanks. LGTM with a comment.

File: src/app-check/index.ts [hunk 6]
```diff
@@ -95,6 +96,18 @@ export namespace appCheck {
...
   }
 
+  /**
+   * Interface representing App Check token options.
+   */
+  export interface AppCheckTokenOptions {
+    /**
+     * The length of time measured in milliseconds starting from when the server
+     * mints the token for which the returned FAC token will be valid.
+     * This value must be in milliseconds and between 30 minutes and 7 days, inclusive.
```
Comment (kevinthecheung) [inline src/app-check/index.ts:106]:
```suggestion
     * The length of time, in milliseconds, for which the App Check token will
     * be valid. This value must be between 30 minutes and 7 days, inclusive.
```

**Review by kevinthecheung — APPROVED** [review]

**Pull request merged**